server/src/main/java/org/opensearch/gateway/AsyncShardBatchFetch.java

```diff
@@ -0,0 +1,341 @@
+ * Cleanup of failed shards is necessary in a batch and based on that a reroute should be triggered to take care of
+ * those in the next run. This separation also takes care of the extra generic type V which is only needed for batch
+ * transport actions like {@link TransportNodesListGatewayStartedShardsBatch}.
+ *
+ * @param <T> Response type of the transport action.
+ * @param <V> Data type of shard level response.
+ *
+ * @opensearch.internal
+ */
+public abstract class AsyncShardBatchFetch<T extends BaseNodeResponse, V> extends AsyncShardFetch<T> {
+
+    private final Consumer<ShardId> removeShardFromBatch;
+    private final List<ShardId> failedShards;
+
+    @SuppressWarnings("unchecked")
+    AsyncShardBatchFetch(
+        Logger logger,
+        String type,
+        Map<ShardId, ShardAttributes> shardAttributesMap,
+        AsyncShardFetch.Lister<? extends BaseNodesResponse<T>, T> action,
+        String batchId,
+        Class<V> clazz,
+        BiFunction<DiscoveryNode, Map<ShardId, V>, T> responseGetter,
+        Function<T, Map<ShardId, V>> shardsBatchDataGetter,
+        Supplier<V> emptyResponseBuilder,
+        Consumer<ShardId> failedShardHandler,
+        Function<V, Exception> getResponseException,
+        Function<V, Boolean> isEmptyResponse
+    ) {
+        super(logger, type, shardAttributesMap, action, batchId);
+        this.removeShardFromBatch = failedShardHandler;
+        this.failedShards = new ArrayList<>();
+        this.cache = new ShardBatchCache<>(
+            logger,
+            type,
+            shardAttributesMap,
+            "BatchID=[" + batchId + "]",
+            clazz,
+            responseGetter,
+            shardsBatchDataGetter,
+            emptyResponseBuilder,
+            this::cleanUpFailedShards,
+            getResponseException,
+            isEmptyResponse
+        );
+    }
+
+    public synchronized FetchResult<T> fetchData(DiscoveryNodes nodes, Map<ShardId, Set<String>> ignoreNodes) {
+        FetchResult<T> result = super.fetchData(nodes, ignoreNodes);
+        if (result.hasData()) {
+            // trigger reroute for failed shards only when all nodes have completed fetching
+            if (failedShards.isEmpty() == false) {
+                // trigger a reroute if there are any shards failed, to make sure they're picked up in next run
+                logger.trace("triggering another reroute for failed shards in {}", reroutingKey);
+                reroute("shards-failed", "shards failed in " + reroutingKey);
+                failedShards.clear();
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Remove the shard from shardAttributesMap so it's not sent in next asyncFetch.
+     * Call removeShardFromBatch method to remove the shardId from the batch object created in
+     * ShardsBatchGatewayAllocator.
+     * Add shardId to failedShards, so it can be used to trigger another reroute as part of upcoming fetchData call.
```

Comment:
Updating to 

```
    /**
     * Remove the shard from shardAttributesMap so it's not sent in next asyncFetch. Because we don't want to do 
     * fetching again in next round for the shards which failed with non retry-able exceptions.
     * Call removeShardFromBatch method to remove the shardId from the batch object created in
     * ShardsBatchGatewayAllocator. So, it's picked up in a new batch in next immediate reroute.
     * Add shardId to failedShards, so it can be used to trigger another reroute as part of upcoming fetchData call.
     *
     * @param shardId shardId to be cleaned up from batch and cache.
     */
```